docs/design-docs/benchmarks.md

```diff
@@ -0,0 +1,274 @@
+
+## High-level Technical Design
+
+![Structure of Benchmark Package](../assets/benchmark-structure.png)
+*Figure 1 Structure of Benchmarks package*
+
+The structure of Benchmarks package can be divided into layers from the bottom up:
+1. Abstract base classes for all kind of benchmarks, including Benchmark, ModelBenchmark, Microbenchmark and DockerBenchmark.
+   1. Benchmark is the base class for all benchmarks. It defines common interfaces such as run(), _preprocess(), _postprocess(), _benchmark(), add_parser_arguments() and so on.
+   2. ModelBenchmark is the base class for all E2E models. It defines the abstract interfaces that need to be implemented by the subclasses using different frameworks, such as PytorchBase, TFBase and ONNXBase. Each subclass will realize part of the abstract interfaces that is common for models, such as _init_distributed_setting(), _init_dataloader(), _create_optimizer().
+   3. Microbenchmark is the base class for all micro benchmarks. It defines the abstract interfaces that need to be implemented by the subclasses, such as _process_raw_result(), _process_numeric_result().
+   4. DockerBenchmark is the base class for real workloads based on docker. It also defines the abstract interfaces that need to be implemented by the subclasses.
+2. Derived classes for all implemented benchmarks, realized all the abstract interfaces. The benchmarks will be registered into Benchmark Registry.
+3. BenchmarkRegistry: provides a way of benchmark registration, maintains all the registered benchmarks and supports benchmark launching by BenchmarkContext.
+4. BenchmarkContext: provides the context to launch one benchmark, including name, parameters, platform(CPU, GPU), framework(Pytorch, TF, ONNX).
+5. BenchmarkResult: define the structured result for each benchmark in json format, including name, return_code, start_time, end_time, raw_data, summarized metrics, etc.
+
+The Executor on the uppermost layer is the entrance for all the benchmarks, it launches the benchmark by BenchmarkRegistry and fetch the BenchmarkResult.
+
+## Detailed Technical Design
+
+This chapter will describe the design details of all the components in Benchmarks package.
+
+### E2E Model Benchmarks
+
+The E2E model benchmarks have 4-layers inheritance relationship.
+
+#### Training 
+
+The general process of model training is:
+
+init_distributed_setting -> generate_dataset -> init_dataloader -> create_model -> create_optimizer -> train
+
+The responsibility for function implementation of every layer is as Figure2. These functions will be executed according to the sequence in the figure. The functions that exist in derived class and not in base class are abstract functions.
+
+![Training process and the responsibility for function implementation of every layer](../assets/model-training-process.png)
+
+*Figure 2 Training process and the responsibility for function implementation of every layer*
```

Comment:
Training process and responsibilities of functions in each layers

Reply:
Do you have some suggestions?

Reply:
I used visio to export the png file again, looks still not clear...

Reply:
Shall we try svg?